tests/lib/openstack_query/test_query_parser.py

```diff
@@ -0,0 +1,415 @@
+import unittest
+from unittest.mock import MagicMock, call, patch
+from parameterized import parameterized
+
+from openstack_query.query_parser import QueryParser
+from tests.lib.openstack_query.mocks.mocked_props import MockProperties
+
+# pylint:disable=protected-access
```

Comment:
I'm tempted to start flagging these, if we're having to do protected access then either we have a hidden inner class doing processing that needs splitting out, or the API needs splitting (like I described with the run pipeline stuff)

Reply:
so we need this because we need to set hidden attributes `_sort_by` and `_group_by`

These are set by the methods `parse_sort_by` and `parse_group_by` 
- called when users build the query using `sort_by`/`group_by` methods
 
Then these attributes are then used in `run_sort_by` and `run_group_by` 
- happens in the run() step 

access to these hidden attributes is needed to test that the "build step" methods actually set the internal attributes properly and also to set the hidden attributes to expected values to test "run step" methods

Reply:
Then along those lines, why can't we mark them as getters and use the public API?

I'm just concerned we're slipping into treating everything as public, as the private / public separation is another way to force people to consider their classes and how they're interacted with

I'll leave this for the moment and circle back to it in the future

Reply:
see #104 for future PR